### PR TITLE
fix: include serial tests in coverage.xml and upload it from CI

### PR DESCRIPTION
**Pull request by runpod-Henrik**
## The bug

Our coverage run is split in two passes:

```make
uv run pytest tests/ ... -m "not serial" --cov=runpod_flash --cov-report=xml
uv run pytest tests/ ... -m "serial"     --cov=runpod_flash --cov-append --cov-report=term-missing
```

The second pass appends to `.coverage` but only re-renders the **terminal** report — it never re-emits the XML. So `coverage.xml` on disk holds the parallel run alone, and the 19 `serial`-marked tests contribute nothing to it.

Measured locally:

| | Coverage | Covered lines |
|---|---:|---:|
| Before the append is re-emitted | 85.1% | 8085 / 9503 |
| After | **85.6%** | **8132 / 9503** |

**47 covered lines were being dropped from the report.** The terminal output was always correct; only the XML was wrong, which is why this went unnoticed — nothing consumed the XML until now.

## Changes

- `Makefile`: add `--cov-report=xml` to the `--cov-append` pass in both `test-coverage` and `ci-quality-github`
- `.github/workflows/ci.yml`: upload `coverage.xml` as `coverage-${{ matrix.python-version }}`

The artifact name is per-matrix-version because `upload-artifact@v4` requires unique names within a run. `if-no-files-found: error` so a silently-missing report fails the step rather than publishing an empty artifact.

## Why upload it

We're automating the weekly **Test Coverage Progress** report, which currently relies on hand-collected numbers. That has produced real errors — figures reported as improvements when they were regressions, and at least one number that matched no CI run at all. With this artifact published, the report reads the number straight from the newest successful run on `main`.

## Verification

Ran the full CI gate (`make ci-quality-github`) in a clean worktree at this commit:

```
53 passed, 1 skipped, 2693 deselected, 2 warnings in 5.47s
Coverage XML written to file coverage.xml
Required test coverage of 65% reached. Total coverage: 85.47%
```

`ruff format --check` (258 files) and `ruff check` both clean. The resulting `coverage.xml` parses to 8122/9503 lines.

No production code touched — CI config and Makefile flags only.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

**Review by deanq — COMMENTED** [review]
A few robustness notes on the coverage plumbing. The core fix (re-emitting coverage.xml after the serial --cov-append pass) is right; these are edge cases around failing/aborted runs.

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -252,13 +252,80 @@ jobs:
...
+              o += ["### Test Results", "",
+                    "| Status | Passed | Failed | Skipped | Total |",
+                    "|---|---:|---:|---:|---:|",
+                    f"| {'✅ Passed' if not f else '❌ Failed'} | {t - f - s} | {f} | {s} | {t} |", ""]
+
+          p = Path(cov)
+          if not p.exists():
+              o += [f"> No coverage report at `{cov}`."]
+          else:
+              r = ET.parse(p).getroot()
```
Comment (deanq) [inline .github/workflows/ci.yml]: The coverage-XML parse here isn't guarded, unlike the junit parse above (the `try/except ET.ParseError` at lines 269-272). Because this step runs under `if: always()`, a truncated or malformed `coverage.xml` — a run killed mid-write by timeout, OOM, or an xdist worker crash — makes `ET.parse(p)` raise, so this summary step exits non-zero and stacks a spurious failure on top of the real one. Consider wrapping this in the same `try/except ET.ParseError` and falling back to the "no coverage report" branch.

File: .github/workflows/ci.yml [hunk 2]
```diff
@@ -252,13 +252,80 @@ jobs:
...
+      # Consumed by the weekly coverage report, which reads the artifact from
+      # the newest successful run on main. Name is per-matrix-version because
+      # upload-artifact@v4 requires unique names within a run.
+      - name: Upload coverage report
+        uses: actions/upload-artifact@v4
+        if: always()
+        with:
+          name: coverage-${{ matrix.python-version }}
+          path: coverage.xml
+          if-no-files-found: error
```
Comment (deanq) [inline .github/workflows/ci.yml]: `if-no-files-found: error` combined with `if: always()` means any run where `coverage.xml` was never produced — e.g. pytest errors at collection/import before pytest-cov writes the file — fails this upload step, red-flagging the job and masking the root cause. On `main` it also stops that run from counting as "successful," which (per the comment above) the weekly coverage report depends on to locate its artifact. Consider `warn` instead of `error` so a missing report doesn't manufacture a second failure or starve the weekly report.

File: Makefile [hunk 3]
```diff
@@ -127,7 +129,9 @@ ci-quality-github: # Quality checks with GitHub Actions formatting (parallel by
 	uv run pytest tests/ --junitxml=pytest-results-parallel.xml -v -n auto -m "not serial" --cov=runpod_flash --cov-report=xml --cov-fail-under=0
 	@echo "::endgroup::"
 	@echo "::group::Test suite with coverage (serial pass)"
-	uv run pytest tests/ --junitxml=pytest-results-serial.xml -v -m "serial" --cov=runpod_flash --cov-append --cov-report=term-missing
+	# Re-emit the XML after appending the serial tests, or coverage.xml is left
+	# holding only the parallel run and undercounts.
+	uv run pytest tests/ --junitxml=pytest-results-serial.xml -v -m "serial" --cov=runpod_flash --cov-append --cov-report=term-missing --cov-report=xml
 	@echo "::endgroup::"
```
Comment (deanq) [inline Makefile:147]: Because the parallel line above (`pytest ... -m "not serial"`) exits non-zero on any test failure, `make` aborts before reaching this serial pass — so on a failing PR `coverage.xml` still holds only the parallel subset and undercounts, which is the exact bug this change is fixing. The re-emit only takes effect when every parallel test passes. If the goal is an accurate number even on failing runs, the parallel invocation needs to not abort the recipe (e.g. a `-` / `|| true` guard) so the serial re-emit always runs. Flagging in case the undercount-on-failure case is the one that matters.

Comment (deanq) [inline Makefile:147]: The parallel invocation passes `--cov-fail-under=0` to suppress the partial-coverage gate, but this serial line doesn't, so it inherits `--cov-fail-under=65` from pyproject `addopts` and gates CI on combined coverage. That's likely intended — the serial pass is where the full number exists — but the asymmetry is undocumented. Worth a short comment noting the serial pass is the real coverage gate, so it isn't later "aligned" to the parallel line and the gate quietly disabled.

File: .github/workflows/ci.yml (hunk 1 above)
Comment (runpod-Henrik) [reply .github/workflows/ci.yml]: Confirmed and fixed. Reproduced it: the old script exits 1 on a truncated `coverage.xml`, the new one exits 0 and reports the parse error in the summary instead.

Your framing sent me looking at the rest of the script and I found the same class of bug in the half I *had* written a guard for — the junit attributes were still going through bare `int()`, so a malformed `tests="abc"` raised too. Both now go through a coercing helper. Also fixed while there: a suite dying at import reports `0 failures` having run nothing, so the status cell rendered a crashed suite as ✅; it now requires that something actually ran.

File: .github/workflows/ci.yml (hunk 2 above)
Comment (runpod-Henrik) [reply .github/workflows/ci.yml]: Changed to `warn` — agreed on the substance. This upload hangs off the PR-gating job, so it shouldn't be able to fail a PR on its own, and it was masking the root cause.

One correction on the second half: a failed upload can't starve the weekly report. If the upload fails there's no artifact either way, and the collector only ever reads runs whose status is `success` — so both paths land on "not re-measured" rather than a wrong number. The masked root cause was the real cost, which is why the change stands.

Kept as `error` on main-ui and RunPod, where this is a dedicated nightly whose only purpose is the artifact.

File: Makefile (hunk 3 above)
Comment (runpod-Henrik) [reply Makefile:147]: Confirmed, and I've documented it — but I'd push back on the remedy. Guarding the parallel line with `-`/`|| true` would make the recipe's exit status come from the serial pass, so a red parallel suite could exit 0. Silently passing a failing suite is worse than an undercount on a run that already failed.

The undercount is also unconsumed: the weekly report only reads runs whose status is `success`, so a red run's `coverage.xml` is never read. The fix does what it needs to on green runs, which are the only ones that matter here. Written into the Makefile so the next reader doesn't have to re-derive it.

Comment (runpod-Henrik) [reply Makefile:147]: Agreed, documented. The comment now states that the serial line *is* the coverage gate and that aligning the two flags would quietly disable it.
